pkg/blockstorage/azure/auth.go

```diff
@@ -0,0 +1,94 @@
+package azure
+
+import (
+	"context"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/kanisterio/kanister/pkg/blockstorage"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// currently avaialble types: https://docs.microsoft.com/en-us/azure/developer/go/azure-sdk-authorization
+// to be available with azidentity: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#readme-credential-types
+// to faciliate future migration, only the common ones between two SDKs are listed
+const (
+	CredTypeManagedIdentity   string = "cred-type-msi"
+	CredTypeClientSecret      string = "cred-type-client-secret"
+	CredTypeClientCertificate string = "cred-type-client-certificate"
+)
+
+var availableCredsType = sets.NewString(CredTypeManagedIdentity)
+
+type CredsValidity int
+
+const (
+	CredValidityUnknown CredsValidity = 0
+	CredsTypeNotSupported
+	CredsFailedAuthentication
+	CredsPassedAuthentication
+)
+
+func isCredTypeSupported(credType string) bool {
+	return availableCredsType.Has(credType)
+}
+
+func isClientCredsAvailable(config map[string]string) bool {
+	if config[blockstorage.AzureTenantID] != "" &&
+		config[blockstorage.AzureCientID] != "" &&
+		config[blockstorage.AzureClentSecret] != "" {
+		return true
+	} else {
+		return false
+	}
+}
+
+func isMSICredsAvailable(config map[string]string) bool {
+	if config[blockstorage.AzureTenantID] == "" &&
+		config[blockstorage.AzureCientID] != "" &&
+		config[blockstorage.AzureClentSecret] == "" {
+		return true
+	} else {
+		return false
+	}
+}
+
+// internal interface to authenticate with different Azure credentials type
+type authenticator interface {
+	authenticate(creds map[string]string) (CredsValidity, error)
```

Comment:
Curious why you need another field of type CredsValidity when the error would have been sufficient ? My recommendation is to create new errors and export them in case the caller wants to look for specific errors/conditions.  With the current model of using CredsValidity, the caller would have to check and use both the validity field and the error field. It is simpler to have the caller deal with only an error field. 
```
var ErrMSIUnsupported = errors.New("MSI Endpoint is not supported")
var ErrServicePrincipalTokenCreateFailed = ...
var ErrRefreshTokenFailed = ....

```

and use them in msIAuthenticator's authenticate . A nil error would indicate that authentication was successful. Only when error is not nil, the caller could check the exported errors from this package.

Reply:
I am not able to think of a design at the moment with other priorities. Between the caller's ability to look for specific errors/conditions and returning errors from Azure APIs, I think it makes sense to prioritize the latter to give end-user precise information on failures. I will : 
1. Remove  `CredsValidity`
2. Keep error returning as is 
Therefore, a successful authentication from the caller side is `error == nil `, any error otherwise will be piped back up the stack

What do you think @onkarbhat

Reply:
Sounds good to me. Thank you Le